README.md
# PortAPI

Easy-peasy, heavily typed, and zod validated requests for browsers, Node, and React Native. Inspired by trpc.

```bash
npm i portapi-helper #npm
yarn add portapi-helper #yarn
pnpm add portapi-helper #pnpm
```

## Basic Usage

```typescript
import { PortAPI } from 'portapi-helper';
import { z } from 'zod';

const client = PortAPI.createClient('https://jsonplaceholder.typicode.com/', {
  headers: {
    'content-type': 'application/json'
  }
});

client.get('todos/1').then((response) => {
  if (!response.success) {
    return console.error(response.error.toString());
  }

  console.log(response.data);
});

// With Zod validation & heavy typing
client
  .get(
    'todos/1',
    z.object({
      id: z.number(),
      title: z.string
    })
  )
  .then((response) => {
    if (!response.success) {
      return console.error(response.error.toString());
    }

    console.log(response.data); // response.data { id: 1, title: 'delectus aut aute' }
  });
```

## Advanced Usage

```typescript
import { PortAPI } from 'portapi-helper';
import { z } from 'zod';

const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {
    headers: {
      'content-type': 'application/json'
    }
  },
  {
    onSuccess: (data) => {},
    onFailedValidation: (error) => {},
    beforeRequest: (request) => {}
    // other handlers
  }
);

client
  .get(
    'todos/1',
    z.object({
      id: z.number(),
      title: z.string()
    })
  )
  .then((response) => {
    if (!response.success) {
      return console.error(response.error.toString());
    }

    console.log(response.data); // response.data { id: 1, title: 'delectus aut aute' }
  });
```

### Request Handlers

Request handlers intercept request at different stages in the request lifecycle.

#### `beforeRequest`

Called before the execution of a request

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    beforeRequest: (request: RequestOptions) => {
      console.log(request.method, request.url, request.body, request.headers);
    }
  }
);
```

#### `onRequest`

Called during the execution of request to mutate the request. Useful for appending the request headers and body.

```typescript
const protectedClient = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onRequest: (request: RequestOptions) => {
      return {
        headers: {
          Authorization: 'Bearer ' + token
        }
      };
    }
  }
);
```

#### `onFailedAuthorization`

Called when a request returns a `403: Unauthorized` status code

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onFailedAuthorization: () => {
      console.log("can't do that!");
    }
  }
);
```

#### `onFailedAuthentication`

Called when a request returns a `401: Unauthenticated` status code

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onFailedAuthentication: () => {
      location.replace('https://www.my-awesome.app/login');
    }
  }
);
```

#### `onFailedRequest`

Called when a request returns a `5xx` status code

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onFailedRequest: (response: any) => {
      // do something
    }
  }
);
```

#### `onFailedParse`

Called when `json.parse()` fails

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onFailedParse: () => {
      // do something
    }
  }
);
```

#### `onFailedValidation`

Called when zod validation fails

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onFailedValidation: (error: z.ZodIssue[]) => {
      console.error(error.join('\n'));
    }
  }
);
```

#### `onSuccess`

Called after successful completion of a request

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onSuccess: (data) => {
      // do something
    }
  }
);
```


CHANGELOG.md
# PortAPI

## 1.0.0

### Major Changes

Easy-peasy, heavily typed, and zod validated requests for browsers, Node, and React Native. Inspired by trpc.

## Basic Usage

```typescript
import { PortAPI } from 'portapi-helper';
import { z } from 'zod';

const client = PortAPI.createClient('https://jsonplaceholder.typicode.com/', {
  headers: {
    'content-type': 'application/json'
  }
});

client.get('todos/1').then((response) => {
  if (!response.success) {
    return console.error(response.error.toString());
  }

  console.log(response.data);
});

// With Zod validation & heavy typing
client
  .get(
    'todos/1',
    z.object({
      id: z.number(),
      title: z.string
    })
  )
  .then((response) => {
    if (!response.success) {
      return console.error(response.error.toString());
    }

    console.log(response.data); // response.data { id: 1, title: 'delectus aut aute' }
  });
```

## Advanced Usage

```typescript
import { PortAPI } from 'portapi-helper';
import { z } from 'zod';

const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {
    headers: {
      'content-type': 'application/json'
    }
  },
  {
    onSuccess: (data) => {},
    onFailedValidation: (error) => {},
    beforeRequest: (request) => {}
    // other handlers
  }
);

client
  .get(
    'todos/1',
    z.object({
      id: z.number(),
      title: z.string()
    })
  )
  .then((response) => {
    if (!response.success) {
      return console.error(response.error.toString());
    }

    console.log(response.data); // response.data { id: 1, title: 'delectus aut aute' }
  });
```

### Request Handlers

Request handlers intercept request at different stages in the request lifecycle.

#### `beforeRequest`

Called before the execution of a request

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    beforeRequest: (request: RequestOptions) => {
      console.log(request.method, request.url, request.body, request.headers);
    }
  }
);
```

#### `onRequest`

Called during the execution of request to mutate the request. Useful for appending the request headers and body.

```typescript
const protectedClient = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onRequest: (request: RequestOptions) => {
      return {
        headers: {
          Authorization: 'Bearer ' + token
        }
      };
    }
  }
);
```

#### `onFailedAuthorization`

Called when a request returns a `403: Unauthorized` status code

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onFailedAuthorization: () => {
      console.log("can't do that!");
    }
  }
);
```

#### `onFailedAuthentication`

Called when a request returns a `401: Unauthenticated` status code

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onFailedAuthentication: () => {
      location.replace('https://www.my-awesome.app/login');
    }
  }
);
```

#### `onFailedRequest`

Called when a request returns a `5xx` status code

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onFailedRequest: (response: any) => {
      // do something
    }
  }
);
```

#### `onFailedParse`

Called when `json.parse()` fails

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onFailedParse: () => {
      // do something
    }
  }
);
```

#### `onFailedValidation`

Called when zod validation fails

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onFailedValidation: (error: z.ZodIssue[]) => {
      console.error(error.join('\n'));
    }
  }
);
```

#### `onSuccess`

Called after successful completion of a request

```typescript
const client = PortAPI.createClient(
  'https://jsonplaceholder.typicode.com/',
  {},
  {
    onSuccess: (data) => {
      // do something
    }
  }
);
```


index.js
import {
    parsePopulate,
    parseTimebox,
    parseFilter,
    parseSearch,
    parseSort,
    parsePaginate
} from "./parsers.js";

export class API {
    constructor(base_url, options = {}) {
        this.base_url = base_url;
        this.options = options;
    }

    async request(method, url, options) {
        let opts = {
            method: method,
            headers: {
                "Content-Type":
                    options.contentType ||
                    this.options.contentType ||
                    "application/json",
                ...this.options.headers,
            },
            ...options,
        };

        if (opts.body) opts.body = JSON.stringify(opts.body);

        try {
            let res = await fetch(this.base_url + url, opts);
            if (res.status >= 200 && res.status < 300)
                return await res.json();
            else return { success: false, ...res };
        } catch (e) {
            console.error(e);
            return e;
        }
    }

    async get(url = "", options = {}) {
        return this.request("GET", url, options);
    }

    async delete(url = "", options = {}) {
        return this.request("DELETE", url, options);
    }

    async post(url = "", body = {}, options = {}) {
        options.body = body;
        return this.request("POST", url, options);
    }

    async patch(url = "", body = {}, options = {}) {
        options.body = body;
        return this.request("PATCH", url, options);
    }

    async query(url = "", query = {}) {
        const { search, filter, sort, paginate, timebox, populate } = query;

        let q = "?"
        q += parsePopulate(populate)
        q += parseTimebox(timebox)
        q += parseFilter(filter)
        q += parseSearch(search)
        q += parseSort(sort)
        q += parsePaginate(paginate)

        q = q.slice(0, -1)

        return this.get(url + q);
    }

    async populate(url = "", populate) {
        return this.query(url, { populate });
    }

    async timebox(url = "", timebox = {}) {
        return this.query(url, { timebox });
    }

    async filter(url = "", filter = {}) {
        return this.query(url, { filter });
    }

    async search(url = "", search = {}) {
        return this.query(url, { search });
    }

    async sort(url = "", sort = {}) {
        return this.query(url, { sort });
    }

    async paginate(url = "", paginate = {}) {
        return this.query(url, { paginate });
    }
}

export default API;

parsers.js
export function parsePopulate(populate) {
    let q = ""
    if (populate) {
        if (Array.isArray(populate.paths))
          populate.paths = populate.paths.join("+");
        q += `populate=${populate.paths}&`;
    }

    return q
}

export function parseTimebox(timebox = {}) {
    let q = ""
    if (timebox.created_after) q += `created_after=${timebox.created_after}&`
    if (timebox.created_before) q += `created_before=${timebox.created_before}&`
    if (timebox.created_prop) q += `created_prop=${timebox.created_prop}&`

    if (timebox.updated_after) q += `updated_after=${timebox.updated_after}&`
    if (timebox.updated_before) q += `updated_before=${timebox.updated_before}&`
    if (timebox.updated_prop) q += `updated_prop=${timebox.updated_prop}&`

    return q
}

export function parseFilter(filter = {}) {
    let q = ""
    if (filter.prop) q += `filter_prop=${filter.prop}&`
    if (filter.value) q += `filter_val=${filter.value}&`
    if (filter.function) q += `filter_fn=${filter.function}&`

    return q
}

export function parseSearch(search = {}) {
    let q = ""
   

    if (search.prop) q += `search_prop=${search.prop}&`
    if (search.value) q += `search_query=${search.value}&`
    if (search.paths) {
        if (Array.isArray(search.paths)) search.paths = search.paths.join("+");
        q += `search_paths=${search.paths}&`;
    }
    if (search.caseSensitive) q += `search_case_sensitive=true&`

    return q
}

export function parseSort(sort = {}) {
    let q = ""
    if (sort.prop) q += `sort_prop=${sort.prop}&`
    if (sort.direction) q += `sort_dir=${sort.direction}&`
    if (sort.function) q += `sort_fn=${sort.function}&`

    return q
}

export function parsePaginate(paginate = {}) {
    let q = ""
    if (paginate.limit) q += `limit=${paginate.limit}&`
    if (paginate.page) q += `page=${paginate.page}&`

    return q
}



src/PortAPI.ts
import { z } from 'zod';
import { PortAPIResponse } from './PortAPIResponse';
import { Method, createPortAPIRequest, methods } from './PortAPIRequest';
import { EventHandlers, EventName } from './PortAPIRequestHandler';

export type IPortAPI = {
  [fn in Lowercase<Method>]:
    | (<T>(
        path: string,
        schema?: z.ZodSchema<T>,
        body?: object
      ) => Promise<PortAPIResponse<T>>)
    | (<T>(
        path: string,
        schema?: z.ZodSchema<T>
      ) => Promise<PortAPIResponse<T>>);
};

export function createClient(
  baseURL: string,
  headers?: Record<string, string>,
  handlers?: Partial<EventHandlers>
) {
  const object: Partial<IPortAPI> = {};

  methods.forEach((method) => {
    if (method === 'POST' || method === 'PUT' || method === 'PATCH')
      object[method.toLowerCase() as Lowercase<Method>] = <T>(
        path: string,
        schema?: z.ZodSchema<T>,
        body?: object
      ) =>
        createPortAPIRequest(
          method,
          `${baseURL}${path}`,
          schema,
          body,
          headers,
          handlers
        ) as Promise<PortAPIResponse<T>>;
    else
      object[method.toLowerCase() as Lowercase<Method>] = <T>(
        path: string,
        schema?: z.ZodSchema<T>
      ) =>
        createPortAPIRequest(
          method,
          `${baseURL}${path}`,
          schema,
          undefined,
          headers,
          handlers
        ) as Promise<PortAPIResponse<T>>;
  });

  return object as IPortAPI;
}


src/PortAPIError.ts
export const portAPIErrors = [
  'INVALID_URL',
  'INVALID_METHOD',
  'INVALID_BODY',
  'INVALID_HEADERS',
  'UNAUTHENTICATED',
  'UNAUTHORIZED',
  'INVALID_RESPONSE',
  'INVALID_STATUS_CODE',
  'INVALID_STATUS_TEXT',
  'INVALID_RESPONSE_JSON',
  'INVALID_RESPONSE_TEXT',
  'INVALID_RESPONSE_BODY',
  'INVALID_RESPONSE_HEADERS',
  'INVALID_RESPONSE_STATUS_CODE'
] as const;

export type PortAPIErrorCode = (typeof portAPIErrors)[number];

export class PortAPIError {
  constructor(
    public type: PortAPIErrorCode,
    public statusCode: number,
    public message?: string,
    public error?: object
  ) {}

  toString() {
    return `${this.type} ${this.statusCode} ${this.message}`;
  }

  toJSON() {
    return {
      type: this.type,
      statusCode: this.statusCode,
      message: this.message,
      error: this.error
    };
  }
}

export const createPortAPIError = (
  type: PortAPIErrorCode,
  statusCode: number,
  message?: string,
  error?: object
) => new PortAPIError(type, statusCode, message, error);


src/PortAPIResponse.ts
import { PortAPIError } from './PortAPIError';

export type PortAPISuccessResponse<T> = {
  success: true;
  data: T;
};

export type PortAPIErrorResponse = {
  success: false;
  error: PortAPIError;
};

type Data<T = any> = T extends PortAPIError ? never : T;

export function createPortAPIResponse<T, R extends boolean>(
  success: R,
  data: R extends true ? T : PortAPIError
): PortAPIResponse<T, R> {
  if (success) {
    return {
      success: true,
      data: data as Data<T>
    } as PortAPIResponse<T, R>;
  }

  return {
    success: false,
    error: data
  } as unknown as PortAPIResponse<T, R>;
}

export type PortAPIResponse<T, R extends boolean = boolean> = R extends true
  ? PortAPISuccessResponse<T>
  : R extends false
  ? PortAPIErrorResponse
  : PortAPISuccessResponse<T> | PortAPIErrorResponse;


src/PortAPIRequest.ts
import fetch from 'cross-fetch';
import { z } from 'zod';
import { createPortAPIError } from './PortAPIError';
import { PortAPIResponse, createPortAPIResponse } from './PortAPIResponse';
import { EventHandlers } from './PortAPIRequestHandler';

export const methods = [
  'GET',
  'POST',
  'PUT',
  'PATCH',
  'DELETE',
  'HEAD',
  'OPTIONS',
  'TRACE'
] as const;
export type Method = (typeof methods)[number];

export async function createPortAPIRequest<T>(
  method: Method,
  url: string,
  schema?: z.ZodSchema<T>,
  body?: any,
  headers: Record<string, string> = {},
  handlers?: Partial<EventHandlers>
) {
  handlers?.beforeRequest?.({ method, url, body, headers });
  if (handlers?.onRequest) {
    const modifiedRequest =
      handlers.onRequest?.({
        method,
        url,
        body,
        headers
      }) ?? {};

    if (modifiedRequest.body)
      body = {
        ...(body ?? {}),
        ...modifiedRequest.body
      };
    if (modifiedRequest.headers)
      headers = {
        ...(headers ?? {}),
        ...modifiedRequest.headers
      };
    if (modifiedRequest.method) method = modifiedRequest.method;
    if (modifiedRequest.url) url += modifiedRequest.url;
  }

  const response = await fetch(url, {
    method,
    body: JSON.stringify(body),
    headers: {
      'Content-Type': 'application/json',
      ...headers
    }
  });

  if (response.status === 401) {
    handlers?.onFailedAuthentication?.();
    return createPortAPIResponse(
      false,
      createPortAPIError('UNAUTHENTICATED', response.status)
    );
  }

  if (response.status === 403) {
    handlers?.onFailedAuthorization?.();
    return createPortAPIResponse(
      false,
      createPortAPIError('UNAUTHENTICATED', response.status)
    );
  }

  if (!response.ok) {
    handlers?.onFailedRequest?.(response);
    return createPortAPIResponse(
      false,
      createPortAPIError('INVALID_RESPONSE', response.status)
    );
  }

  const json = (await response.json().catch(() => {
    handlers?.onFailedParse?.();
    return createPortAPIResponse(
      false,
      createPortAPIError('INVALID_RESPONSE_JSON', response.status)
    );
  })) as T;

  if (schema) {
    const validationResult = schema.safeParse(json);
    if (!validationResult.success) {
      handlers?.onFailedValidation?.(validationResult.error.errors);
      return createPortAPIResponse(
        false,
        createPortAPIError(
          'INVALID_RESPONSE_JSON',
          response.status,
          validationResult.error.message,
          validationResult.error
        )
      );
    }
    handlers?.onSuccess?.(validationResult.data);
    return createPortAPIResponse(true, validationResult.data);
  }

  handlers?.onSuccess?.(json);
  return createPortAPIResponse(true, json);
}

export type PortAPIRequest<T> = (
  method: Method,
  url: string,
  schema?: z.ZodSchema<T>,
  body?: any,
  headers?: Record<string, string>
) => Promise<PortAPIResponse<T>>;

export type RequestOptions = {
  method: Method;
  url: string;
  body?: any;
  headers?: Record<string, string>;
};
